messaging/src/main/java/org/springframework/security/messaging/context/SecurityContextChannelInterceptor.java

```diff
@@ -91,7 +98,18 @@ public void setAnonymousAuthentication(Authentication authentication) {
 
 	@Override
 	public Message<?> preSend(Message<?> message, MessageChannel channel) {
-		setup(message);
```

Comment:
It looks like what this previously did was set the security context to whatever user was in the message (anonymous if no user).

With the new logic, that won't happen anymore. Instead, if there is no user, then an authenticated security context will remain as-is.

Also, it seems that if it is an authenticated security context, then it doesn't get pushed to the stack. This creates problems if `cleanup` is called later on. For example, if the message header is empty and `SecurityContextHolder` is set to a value, when `afterSendCompletion` is done, the security context is emptied instead of set back to what it what before.

To remain passive, I think that `setup` needs to be called unconditionally.

Would it make sense to do the following instead:

```java
setup(message);
if (message.getHeaders().get(this.authenticationHeaderName) == null) {
    Authentication authentication = this.securityContextHolderStrategy.getContext().getAuthentication();
    message = MessageBuilder.fromMessage(message).setHeader(this.authenticationHeaderName, authentication).build();
}
```

In this way, the `SecurityContextHolder` stack is preserved and the message gets the current security context as a value (anonymous if there is no security context).

Reply:
And that is exactly not OK for the change I'm proposing.
The `setup(message);` does set an `anonymous` if no header in the message, but my goal is to take the current one and set it into header.

Perhaps `cleanup()` must not pop if we haven't pushed...

The goal of this change is to add a new behavior to populate the header from the current context not changing it and not pushing anything to the stack. If header is present and it is not the same what we have on a context, then yes we do push.

The previous version if this interceptor is indeed what you are describing: push if header present or not and this does not give us what we need on the receiving side. Perhaps it will work if we would add that header in the other interceptor upfront this one, but why if there is already one in the current `SecurityContext`.

Let me see what we can do with `cleanup()` to follow this new logic!

Reply:
> The setup(message); does set an anonymous if no header in the message, but my goal is to take the current one and set it into header.

It makes me wonder if the logic you are wanting to have should be in a separate interceptor. This interceptor wasn't originally intended as propagation but as an analogy to `SecurityContextHolderFilter`. It can't behave in opposing ways based on the same input (no header).

Reply:
Sure! I also thought about that.
Currently the `SecurityContextChannelInterceptor` really does propagate the ctx via headers if the value is present in the message.
To avoid duplication an interceptor you suggest and I've thought for would only load the current ctx into a header and this `SecurityContextChannelInterceptor` would be used as a second one to propagate the ctx.
And that is going to be a typical configuration.
This way end-users may end up with the question why do we need two interceptors when only one would do the trick for both use-cases.

If you still think that `SecurityContextChannelInterceptor` must remain as it is, then I'll work on a  `SecurityContextPropagationChannelInterceptor`  even if it is going to have some logic similar to the `SecurityContextChannelInterceptor`.
The only different with this two interceptors would be only in a way how they deal with current ctx and header: one of them takes header and set it into a current ctx.
Another takes current ctx in a `preSend` and populate int on other threads in the `postReceive()` and `beforeHandle()`.
Well, probably `afterSendCompletion()` won't do anything in this new interceptor, since we really don't affect the current ctx, but just takes it into the message header.

Let me know your decision and I'll proceed from here!

Thank you for feedback!

Reply:
The existing interceptor is designed to populate the security context with the header; this PR wants to populate the header with the security context. While there is some overlap, I think these two goals are ultimately at odds. One is (conceptually) `MessageToSecurityContextInterceptor` and the other is `SecurityContextToMessageInterceptor`.

> And that is going to be a typical configuration.

It seems to me that you'd want to populate the security context (`SecurityContextChannelInterceptor`) on the inbound channel and populate the message (this PR) on the outbound channel. So when we talk about typical configuration, I am doubting that they would be in the same interceptor chain. Indeed, it seems like they'd cancel out each other just a bit.

If it sounds like I'm understanding correctly, then yes please, let's move this functionality into a separate filter. I wonder if `SecurityContextOutboundChannelInterceptor` would be an appropriate name. Since my recommendation is based on my beginner-level understanding of message channels, please feel free to help me see things differently.

Reply:
Thanks, Josh, for confirmation!

I'm looking not into a separate `SecurityContextPropagationChannelInterceptor` impl.

To clean things a little:
No, we don't talk about channel adapters. 
We talk about a thread boundaries provided by the `ExecutorChannel` (with some variations), where `send` operation happens on one thread and then `handle` on the other. That's exactly why we have that special `ExecutorChannelInterceptor` abstraction. 
Another one is a `QueueChannel` where `send` puts the message onto the queue and `receive` pulls it from there, which typically happens on a scheduled thread.
Both of those threads just two separate ends of the same channel, therefore we have a scope of the message as a unit to carry all the information via interceptor hooks: https://www.enterpriseintegrationpatterns.com/patterns/messaging/MessageChannel.html